enterprise/cmd/worker/internal/codeintel/indexing/dependency_sync_scheduler.go

```diff
@@ -0,0 +1,204 @@
+	dbStore     DBStore
+	workerStore dbworkerstore.Store
+	extsvcStore ExternalServiceStore
+}
+
+func (h *dependencySyncSchedulerHandler) Handle(ctx context.Context, record workerutil.Record) error {
+	if !autoIndexingEnabled() {
+		return nil
+	}
+
+	job := record.(dbstore.DependencySyncingJob)
+
+	scanner, err := h.dbStore.ReferencesForUpload(ctx, job.UploadID)
+	if err != nil {
+		return errors.Wrap(err, "dbstore.ReferencesForUpload")
+	}
+	defer func() {
+		if closeErr := scanner.Close(); closeErr != nil {
+			err = multierror.Append(err, errors.Wrap(closeErr, "dbstore.ReferencesForUpload.Close"))
+		}
+	}()
+
+	var (
+		kinds                      = map[string]struct{}{}
+		oldDependencyReposInserted int
+		newDependencyReposInserted int
+		errs                       []error
+	)
+
+	for {
+		packageReference, exists, err := scanner.Next()
+		if err != nil {
+			return errors.Wrap(err, "dbstore.ReferencesForUpload.Next")
+		}
+		if !exists {
+			break
+		}
+
+		pkg := precise.Package{
+			Scheme:  packageReference.Package.Scheme,
+			Name:    packageReference.Package.Name,
+			Version: packageReference.Package.Version,
+		}
+
+		extsvcKind, ok := schemeToExternalService[packageReference.Scheme]
+		// add entry for empty string/kind here so dependencies such as lsif-go ones still get
+		// an associated dependency indexing job
+		kinds[extsvcKind] = struct{}{}
+		if !ok {
+			continue
+		}
+
+		new, err := h.insertDependencyRepo(ctx, pkg)
+		if err != nil {
+			errs = append(errs, err)
+		} else if new {
+			newDependencyReposInserted++
+		} else {
+			oldDependencyReposInserted++
+		}
+	}
+
+	var nextSync time.Time
+	// If len == 0, it will return all external services, which we definitely don't want.
+	if len(kindsToArray(kinds)) > 0 {
+		nextSync = time.Now()
+		externalServices, err := h.extsvcStore.List(ctx, database.ExternalServicesListOptions{
+			Kinds: kindsToArray(kinds),
+		})
+		if err != nil {
+			if len(errs) == 0 {
+				return errors.Wrap(err, "dbstore.List")
+			} else {
+				return multierror.Append(err, errs...)
+			}
+		}
+
+		log15.Info("syncing external services",
+			"upload", job.UploadID, "numExtSvc", len(externalServices), "job", job.ID, "schemaKinds", kinds,
+			"newRepos", newDependencyReposInserted, "existingInserts", oldDependencyReposInserted)
+
+		for _, externalService := range externalServices {
+			externalService.NextSyncAt = nextSync
+			err := h.extsvcStore.Upsert(ctx, externalService)
+			if err != nil {
+				errs = append(errs, errors.Wrapf(err, "extsvcStore.Upsert: error setting next_sync_at for external service %d - %s", externalService.ID, externalService.DisplayName))
+			}
+		}
+	} else {
+		log15.Info("no package schema kinds to sync external services for", "upload", job.UploadID, "job", job.ID)
+	}
+
+	shouldIndex, err := h.shouldIndexDependencies(ctx, h.dbStore, job.UploadID)
+	if err != nil {
+		return err
+	}
+
+	if shouldIndex {
+		// see note on L94 on how this is populated
```

Comment:
```suggestion
		// If we saw a kind that's not in schemeToExternalService, then kinds contains an empty string key
```

oof direct line indexes will be super brittle.